### DISCUSSION TITLE
How to save homography matrix (for video)

### DISCUSSION THREAD

**Discussion by HugoDeCozar** [OpenStitching/stitching#148] [Q&A] [answered]
Hello!

I am stitching frames from two videos (left and right camera) to make a unique video. Stitching individual frames is fine, but when the resulting frames are combined in the final video, the result is very messy. This is because the images will warp a lot from one frame to the other.

Ill post two frames as an example

![1](https://github.com/OpenStitching/stitching/assets/145971234/4cd48c8b-3c29-4b9c-8ad3-65288a2647aa)
![2](https://github.com/OpenStitching/stitching/assets/145971234/f4604ee0-af2e-42a3-9b57-df74897ccf64)

I guess the solution is to get the homography matrix from stitching the first frame, and using it in the other frames. I got that idea from this tutorial: https://pyimagesearch.com/2016/01/25/real-time-panorama-and-image-stitching-with-opencv/

How would you do that?

**Comment by lukasalexanderweber** [comment, marked as answer]
You could try something like this (not tested!):

```python
from stitching import Stitcher
from stitching.images import Images

class VideoStitcher(Stitcher):

    def initialize_stitcher(self, **kwargs):
        super().initialize_stitcher(kwargs)
        self.cameras = None
        self.cameras_registered = False
		
    def stitch(self, images, feature_masks=[]):
        self.images = Images.of(
            images, self.medium_megapix, self.low_megapix, self.final_megapix
        )

        if not self.cameras_registered:
            imgs = self.resize_medium_resolution()
            features = self.find_features(imgs, feature_masks)
            matches = self.match_features(features)
            imgs, features, matches = self.subset(imgs, features, matches)
            cameras = self.estimate_camera_parameters(features, matches)
            cameras = self.refine_camera_parameters(features, matches, cameras)
            cameras = self.perform_wave_correction(cameras)
            self.estimate_scale(cameras)
            self.cameras = cameras
            self.cameras_registered = True

        imgs = self.resize_low_resolution(imgs)
        imgs, masks, corners, sizes = self.warp_low_resolution(imgs, self.cameras)
        self.prepare_cropper(imgs, masks, corners, sizes)
        imgs, masks, corners, sizes = self.crop_low_resolution(
            imgs, masks, corners, sizes
        )
        self.estimate_exposure_errors(corners, imgs, masks)
        seam_masks = self.find_seam_masks(imgs, corners, masks)

        imgs = self.resize_final_resolution()
        imgs, masks, corners, sizes = self.warp_final_resolution(imgs, self.cameras)
        imgs, masks, corners, sizes = self.crop_final_resolution(
            imgs, masks, corners, sizes
        )
        self.set_masks(masks)
        imgs = self.compensate_exposure_errors(corners, imgs)
        seam_masks = self.resize_seam_masks(seam_masks)

        self.initialize_composition(corners, sizes)
        self.blend_images(imgs, seam_masks, corners)
        return self.create_final_panorama()
```

So basically you get the cameras with your first image set and afterwards use the obtained cameras for all other frames

**Reply by earthscience2** [reply]
The three images were stitched using the corresponding code. Although the camera variables were stitched while fixed, a small part of the screen continued to change in the created image. Is it normal?

**Reply by lukasalexanderweber** [reply]
No

**Comment by HugoDeCozar** [comment]
Works great! Thanks a lot

The only change i had to make is placing _imgs = self.resize_medium_resolution()_ ABOVE _if not self.cameras_registered_.

**Reply by lukasalexanderweber** [reply]
Great! I think changing 

`imgs = self.resize_low_resolution(imgs)`
to 
`imgs = self.resize_low_resolution()`

should also work

**Comment by HugoDeCozar** [comment]
Another thing: my frames now take 0.2s each to stitch. Is there a way to make the process faster? I already have _try_use_gpu=True_.

**Reply by lukasalexanderweber** [reply]
> Great! I think changing
> 
> `imgs = self.resize_low_resolution(imgs)` to `imgs = self.resize_low_resolution()`
> 
> should also work

I think this is faster, since the images don't need to be resized twice.

If you don't need the full resolution you can reduce it by setting `final_megapix`.

You maybe don't need exposure compensation, seam finding or blending, depending on your requirements on the result 
`compensator="no", blender_type="no", finder="no"`

**Reply by HugoDeCozar** [reply]
thanks! will try

**Comment by BALLABLLU** [comment]
Can I use a GPU to increase the frame rate? For now, I want to use a while loop and a buffer to achieve real-time panorama stitching.

**Reply by lukasalexanderweber** [reply]
I'm by far not an expert in GPU acceleration, and many people already asked in the forum If its possible with stitching. I was never able to test CUDA support since I don't have NVIDIA hardware.

**Comment by BALLABLLU** [comment]
Thank you for answering me

Lukas Weber ***@***.***>于2023年10月9日 周一上午1:26写道：

**Reply by deepConnectionism** [reply]
I also encountered this problem. How can I stitch images from 8 1080P cameras in real time? Need to achieve 30 fps

**Comment by iiiceice** [comment]
> Hello!
> 
> I am stitching frames from two videos (left and right camera) to make a unique video. Stitching individual frames is fine, but when the resulting frames are combined in the final video, the result is very messy. This is because the images will warp a lot from one frame to the other.
> 
> Ill post two frames as an example
> 
> ![1](https://private-user-images.githubusercontent.com/145971234/271342033-4cd48c8b-3c29-4b9c-8ad3-65288a2647aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg1NjI5MTUsIm5iZiI6MTcyODU2MjYxNSwicGF0aCI6Ii8xNDU5NzEyMzQvMjcxMzQyMDMzLTRjZDQ4YzhiLTNjMjktNGI5Yy04YWQzLTY1Mjg4YTI2NDdhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMFQxMjE2NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNjljYWEyZTE4ODQ3Y2YwODY3MjhjMWVjOTgzMzk3YTA2ODVkYWM2Y2ZjNDU4ZTdhOWUyYzVhMTFmMGIyYjhhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.J4l6Tp__pNiwYw9vd-zmBZS8Uv6q9qxHLI765368VC8) ![2](https://private-user-images.githubusercontent.com/145971234/271342062-f4604ee0-af2e-42a3-9b57-df74897ccf64.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg1NjI5MTUsIm5iZiI6MTcyODU2MjYxNSwicGF0aCI6Ii8xNDU5NzEyMzQvMjcxMzQyMDYyLWY0NjA0ZWUwLWFmMmUtNDJhMy05YjU3LWRmNzQ4OTdjY2Y2NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMFQxMjE2NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMjUyNmI4MWVkOTFmYjJkMmM4ZjE0MmNhMGZkOTgxMDNjMjkxNmE0M2E1ZGQwMDViNTE0ZDkxMGM1MjE2MzhkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Zt3b58_RAxEAMPjdD5iesg8u0WUkcB0Hf5NQFG3iLds)
> 
> I guess the solution is to get the homography matrix from stitching the first frame, and using it in the other frames. I got that idea from this tutorial: https://pyimagesearch.com/2016/01/25/real-time-panorama-and-image-stitching-with-opencv/
> 
> How would you do that?

can you share that original code? cause i found that link is to subscribe, and that costs much, i can't afford that.PLEASE!

**Comment by mikeuppp** [comment]
> Hello!
> 
> I am stitching frames from two videos (left and right camera) to make a unique video. Stitching individual frames is fine, but when the resulting frames are combined in the final video, the result is very messy. This is because the images will warp a lot from one frame to the other.
> 
> Ill post two frames as an example
> 
> ![1](https://private-user-images.githubusercontent.com/145971234/271342033-4cd48c8b-3c29-4b9c-8ad3-65288a2647aa.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzkyOTkzNTMsIm5iZiI6MTc3OTI5OTA1MywicGF0aCI6Ii8xNDU5NzEyMzQvMjcxMzQyMDMzLTRjZDQ4YzhiLTNjMjktNGI5Yy04YWQzLTY1Mjg4YTI2NDdhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDUyMFQxNzQ0MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNzdiYjFjZTRkMDhkMzUzNmViM2M2YjRlYzMyMWYyYTkxNDZjNjc3NjU5ZDU4YWVhOTJlOTUzZGY0MGQ5OTQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.SZO_Csqy06hpTqvp7SGO7qhWO0l7W5_eHFqIt3kFbBg) ![2](https://private-user-images.githubusercontent.com/145971234/271342062-f4604ee0-af2e-42a3-9b57-df74897ccf64.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzkyOTkzNTMsIm5iZiI6MTc3OTI5OTA1MywicGF0aCI6Ii8xNDU5NzEyMzQvMjcxMzQyMDYyLWY0NjA0ZWUwLWFmMmUtNDJhMy05YjU3LWRmNzQ4OTdjY2Y2NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDUyMFQxNzQ0MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZWY4MmQ5YWE4MTkyMDBiOWEwYjE2YWJmODNkYjlkMDg3NmVhNTQ3NzZlZjAxZjc5YTgwMjdhNjcyZjU1NjY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.9RqTaUhhf-DWfvXnBcYDaLWhQ5uufcuMZ2dU-lVZrcM)
> 
> I guess the solution is to get the homography matrix from stitching the first frame, and using it in the other frames. I got that idea from this tutorial: https://pyimagesearch.com/2016/01/25/real-time-panorama-and-image-stitching-with-opencv/
> 
> How would you do that?

Hi.
Nice to meet you.
I need your help. I will compensate for that.
Shall we discuss more detail?
